nncf/common/quantization/structs.py

```diff
@@ -0,0 +1,175 @@
+        if self.num_bits > other.num_bits:
+            return False
+        if self.mode is QuantizationMode.ASYMMETRIC and other.mode is \
+                QuantizationMode.SYMMETRIC:
+            return False
+        if self.signedness_to_force is None and other.signedness_to_force is not None:
+            return False
+        if self.signedness_to_force is True and other.signedness_to_force is False:
+            return False
+
+        return True
+
+    def compatible_with_a_unified_scale_linked_qconfig(self, linked_qconfig: 'QuantizerConfig'):
+        return self.num_bits == linked_qconfig.num_bits and \
+               self.mode == linked_qconfig.mode and \
+               self.signedness_to_force == linked_qconfig.signedness_to_force and \
+               self.per_channel == linked_qconfig.per_channel
+
+    def is_a_bitwidth_variant(self, other_qconfig: 'QuantizerConfig') -> bool:
+        return self.per_channel == other_qconfig.per_channel and \
+               self.signedness_to_force == other_qconfig.signedness_to_force and \
+               self.mode == other_qconfig.mode
+
+
+class QuantizerSpec:
+    """A specific (potentially framework-aware) parameter struct required to initialize a
+    given object that performs quantization of an input tensor."""
+    def __init__(self, num_bits: int,
+                 mode: QuantizationMode,
+                 signedness_to_force: bool,
+                 narrow_range: bool):
+        self.num_bits = num_bits
+        self.mode = mode
+        self.signedness_to_force = signedness_to_force
+        self.narrow_range = narrow_range
+
+    def __eq__(self, other: 'QuantizerSpec'):
+        return self.__dict__ == other.__dict__
+
+    @classmethod
+    def from_config(cls, qconfig: QuantizerConfig, narrow_range: bool):
+        return cls(qconfig.num_bits,
+                   qconfig.mode,
+                   qconfig.signedness_to_force,
+                   narrow_range)
+
+
+class QuantizerSetupType(Enum):
+    PATTERN_BASED = "pattern_based"
+    PROPAGATION_BASED = "propagation_based"
+
+    @staticmethod
+    def from_str(quantizer_setup_type: str) -> 'QuantizerSetupType':
+        if quantizer_setup_type == QuantizerSetupType.PATTERN_BASED.value:
+            return QuantizerSetupType.PATTERN_BASED
+        if quantizer_setup_type == QuantizerSetupType.PROPAGATION_BASED.value:
+            return QuantizerSetupType.PROPAGATION_BASED
+        raise RuntimeError("Unknown quantizer setup type. Please select 'pattern_based' or 'propagation_based'.")
+
+
+class QuantizationConstraints:
+    REF_QCONF_OBJ = QuantizerConfig()
+
+    def __init__(self, **kwargs):
+        """Use attribute names of QuantizerConfig as arguments
+        to set up constraints.
+        E.g. QuantizationConstraint(bits=8, per_channel=True) will set up
+        a constraint that corresponds to all 8-bit per-channel quantizers, either
+        symmetric or asymmetric, either signed or unsigned."""
```

Comment:
Please, make your changes conform to Python Style Guide

Reply:
Done